### ISSUE TITLE
TypeError: Cannot read property 'length' of undefined at getAvailabilityZones

### ISSUE THREAD

**Issue by damianh** [pulumi/pulumi#3748]
A stack I've been working with has stopped working for `preview` and `up` for me.

This code: 

```
import * as aws from "@pulumi/aws";
import * as awsx from "@pulumi/awsx";

const provider = new aws.Provider(
    "my-provider",
    {
        sharedCredentialsFile: "~/.aws/sandbox-credentials"
    });

const myVpc = new awsx.ec2.Vpc("my-vpc", {
    subnets: [
        {type: "public"},
        {type: "private"},
    ],
    numberOfNatGateways: 1,
}, { provider });
```

`pulumi preview` results in:

```
Previewing update (dev):

     Type                     Name                Plan       Info
     pulumi:pulumi:Stack      aws-typescript-dev             1 error
 +   ├─ awsx:x:ec2:Vpc        my-vpc              create
 +   └─ pulumi:providers:aws  sandbox-provider    create
 
Diagnostics:
  pulumi:pulumi:Stack (aws-typescript-dev):
    error: Running program 'D:\dev\damianh\pulumi-vpc\src' failed with an unhandled exception:
    TypeError: Cannot read property 'length' of undefined
        at getAvailabilityZones (D:\dev\damianh\pulumi-vpc\src\node_modules\@pulumi\ec2\vpc.ts:302:22)
        at new Vpc (D:\dev\damianh\pulumi-vpc\src\node_modules\@pulumi\ec2\vpc.ts:63:39)
        at Object.<anonymous> (D:\dev\damianh\pulumi-vpc\src\index.ts:12:15)
        at Module._compile (internal/modules/cjs/loader.js:955:30)
        at Module.m._compile (D:\dev\damianh\pulumi-vpc\src\node_modules\ts-node\src\index.ts:814:23)
        at Module._extensions..js (internal/modules/cjs/loader.js:991:10)
        at Object.require.extensions.<computed> [as .ts] (D:\dev\damianh\pulumi-vpc\src\node_modules\ts-node\src\index.ts:817:12)
        at Module.load (internal/modules/cjs/loader.js:811:32)
        at Function.Module._load (internal/modules/cjs/loader.js:723:14)
        at Module.require (internal/modules/cjs/loader.js:848:19)
```

Node version: v12.14.1
Pulumi version: v1.8.1

I had to set this to prevent the hanging issue:
`pulumi:noSyncCalls: "true"`

I've posted a repro project to https://github.com/damianh/pulumi-vpc
I've validated this with a colleague.

I'm dead in the water with this :(

**Comment by clstokes** [comment]
I'm able to reproduce this using Damian's referenced repro project. On my machine, removing the `pulumi:noSyncCalls: "true` from the stack config resolves the issue.

**Comment by CyrusNajmabadi** [comment]
investigating.

**Comment by damianh** [comment]
@clstokes What version of node are you using?

If I remove that, then `pulumi preview` hangs.

**Comment by clstokes** [comment]
I'm using `v10.16.3`.

**Comment by CyrusNajmabadi** [comment]
Reproes for me as well.  Looking further.

**Comment by damianh** [comment]
@clstokes If you use latest Node LTS version you'll probably be hit with #3309 
@CyrusNajmabadi relieved to hear :)

**Comment by CyrusNajmabadi** [comment]
Ok.  So this is fixed in pulumi-awsx 0.18.15.  But that's in preview right now.  Working on figuring out a way forward for you.

**Comment by damianh** [comment]
Confirmed that the issue does not occur with `@pulumi/awsx 0.18.15-alpha.1578507252`. To unblock myself, I can take a reference on this package pending a release.

Thanks for the fast investigation.

**Comment by CyrusNajmabadi** [comment]
Hi @damianh So i don't have great news for you.  The problem *right now* is the use of `pulumi:noSyncCalls: "true"` and this version of the library.  They're just not compatible together.  TO move forward (until we release 0.18.15) you'll need to remove `pulumi:noSyncCalls: "true"`.

Now, as you mentioned, this gives you a hang.  The path forward on that, unfortunately, is to take the approach specified here: https://www.pulumi.com/docs/troubleshooting/#register-the-provider-first

You'll specifically have to do this:

```ts
const provider = new aws.Provider(
    "my-provider",
    {
        sharedCredentialsFile: "~/.aws/sandbox-credentials"
    });

// this line is new
await ProviderResource.register(provider);

const myVpc = new awsx.ec2.Vpc("my-vpc", {
    subnets: [
        {type: "public"},
        {type: "private"},
    ],
    numberOfNatGateways: 1,
}, { provider });
```

Note: because this uses `await`,  you'll need to use `async/await`.  you could do this like so:

```ts
import * as pulumi from "@pulumi/pulumi";
import * as aws from "@pulumi/aws";
import * as awsx from "@pulumi/awsx";

export = async () => {
    const provider = new aws.Provider(
        "my-provider",
        {
            sharedCredentialsFile: "~/.aws/sandbox-credentials",
            region: "us-east-2"
        });

    await pulumi.ProviderResource.register(provider);

    const myVpc = new awsx.ec2.Vpc("my-vpc", {
        subnets: [
            {type: "public"},
            {type: "private"},
        ],
        numberOfNatGateways: 1,
    }, { provider });
}
```

I have tested this out and validated that it works:

![image](https://user-images.githubusercontent.com/4564579/72384355-24ffd700-36d2-11ea-8f4c-a9bc541f5c73.png)

**Comment by damianh** [comment]
Great! I will go ahead with your suggestion. Cheers.

**Comment by seivan** [comment]
@CyrusNajmabadi What is the purpose of 

`ProviderResource.register(provider);` 

and still setting the provider with `, { provider });`
?

**Comment by CyrusNajmabadi** [comment]
@seivan Hooboy, it's long and complex to explain.  Here's the best way i can sum it up. 

When you pass `{ provider }` to a resource, we end up using it to inform queries we make (like `getAvailabilityZones`).  i.e. we're using that provider to answer the question `what region are you asking for availability zones in?`.

This currently has some significant problems because we end up making that call to `getAvailabilityZones` synchronously and we depend on a library to do that that ends up causing hangs and other issues for many users (across many versins of nodejs and OS).

To avoid this problem *currently* the use of `await pulumi.ProviderResource.register(provider);` allows us to avoid the codepath that goes through that problematic library.  i.e. without that line, we end up calling `problematiclibrary.uhoh(()=> ProviderResource.register(provider)`. 

By pulling this line of code out to a location that *can* be `async`, we side-step this issue until the full fix and updates to all our libraries are released.

**Comment by seivan** [comment]
@CyrusNajmabadi Thanks a bunch for explaining that, appreciate it.

**Comment by lukehoban** [comment]
I think this issue is technically "fixed" already - but we can close out once `0.18.5` is released.

**Comment by lukehoban** [comment]
This is now available in https://www.npmjs.com/package/@pulumi/awsx/v/0.19.0.
